### PR TITLE
Add issue count to daily docket

### PR DESCRIPTION
**Pull request by sharonwarner**
Connects #3848 

- [ ] Added unit test to make sure count works across multiple appeal streams
- [ ] Confirmed it looks good in the UI
- [ ] Ran the query in UAT

screen shot 2017-12-18 at 4 40 34 pm

### CODE REVIEW

File: app/repositories/hearing_repository.rb [hunk 1]
```diff
@@ -8,7 +8,19 @@ def upcoming_hearings_for_judge(css_id)
         VACOLS::CaseHearing.upcoming_for_judge(css_id) +
           VACOLS::TravelBoardSchedule.upcoming_for_judge(css_id)
       end
-      hearings_for(MasterRecordHelper.remove_master_records_with_children(records))
+      hearings = hearings_for(MasterRecordHelper.remove_master_records_with_children(records))
+      hearings_appeals = hearings.select { |h| h.master_record == false }.map(&:appeal)
```
Comment (aroltsch) [inline app/repositories/hearing_repository.rb]: Each Vacols::CaseHearing already has an appeal id (bfkey?), you can set the appeal_vacols_id in here https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/repositories/hearing_repository.rb#L91 and then remove this method https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/models/hearing.rb#L161

Comment (aroltsch) [inline app/repositories/hearing_repository.rb]: Each Vacols::CaseHearing already has an appeal id (bfkey?), you can set the appeal_vacols_id in here https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/repositories/hearing_repository.rb#L91 and then remove this method https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/models/hearing.rb#L161

File: app/repositories/hearing_repository.rb [hunk 2]
```diff
@@ -8,7 +8,19 @@ def upcoming_hearings_for_judge(css_id)
...
+      # To speed up the daily docket and the hearing worksheet page loads, we pull in issues for appeals here.
+      hearings_appeals_issues = VACOLS::CaseIssue.descriptions(hearings_appeals.map(&:vacols_id))
+      hearings_appeals.map do |hearing_appeal|
+        appeal_issues_hash_array = hearings_appeals_issues[hearing_appeal.vacols_id] || []
+        appeal = Appeal.find_or_initialize_by(vacols_id: hearing_appeal.vacols_id)
```
Comment (aroltsch) [inline app/repositories/hearing_repository.rb]: I thought that `hearing_appeal` is already an Appeal object. On line 12, you are pulling appeals. However, if you make the changes I suggested above, then you will need to find an appeal in the DB.

File: app/repositories/hearing_repository.rb [hunk 3]
```diff
@@ -8,7 +8,19 @@ def upcoming_hearings_for_judge(css_id)
         VACOLS::CaseHearing.upcoming_for_judge(css_id) +
           VACOLS::TravelBoardSchedule.upcoming_for_judge(css_id)
       end
-      hearings_for(MasterRecordHelper.remove_master_records_with_children(records))
+      hearings = hearings_for(MasterRecordHelper.remove_master_records_with_children(records))
+      hearings_appeals = hearings.select { |h| h.master_record == false }.map(&:appeal)
+      # To speed up the daily docket and the hearing worksheet page loads, we pull in issues for appeals here.
+      hearings_appeals_issues = VACOLS::CaseIssue.descriptions(hearings_appeals.map(&:vacols_id))
```
Comment (aroltsch) [inline app/repositories/hearing_repository.rb]: why don't you just call it `issues` :) ?

File: app/repositories/hearing_repository.rb [hunk 4]
```diff
@@ -8,7 +8,19 @@ def upcoming_hearings_for_judge(css_id)
...
+      hearings_appeals.map do |hearing_appeal|
+        appeal_issues_hash_array = hearings_appeals_issues[hearing_appeal.vacols_id] || []
+        appeal = Appeal.find_or_initialize_by(vacols_id: hearing_appeal.vacols_id)
+        if appeal.worksheet_issues.empty? do
```
Comment (aroltsch) [inline app/repositories/hearing_repository.rb]: I don't think you need `do` for the `if` statement.

File: app/repositories/hearing_repository.rb [hunk 5]
```diff
@@ -8,7 +8,19 @@ def upcoming_hearings_for_judge(css_id)
...
+             appeal_issues_hash_array.map { |i| WorksheetIssue.create_from_issue(appeal, i) }
+           end
+        end
+      end
```
Comment (aroltsch) [inline app/repositories/hearing_repository.rb]: I would also recommend putting this logic into a separate method or even methods so we can re-use the code.

File: app/repositories/hearing_repository.rb (hunk 3 above)
Comment (sharonwarner) [reply app/repositories/hearing_repository.rb]: 😂 

I was mimicking the code that Reader uses to do the same thing, and I didn't even think about updating the variable names...

File: app/repositories/hearing_repository.rb [hunk 6]
```diff
@@ -8,7 +8,22 @@ def upcoming_hearings_for_judge(css_id)
...
+      # To speed up the daily docket and the hearing worksheet page loads, we pull in issues for appeals here.
+      Hearing.repository.load_issues(non_master_record_hearings)
+      hearings
+    end
+
+    def load_issues(hearings)
+      issues = VACOLS::CaseIssue.descriptions(hearings.map(&:appeal_vacols_id))
+      hearings.map do |hearing|
+        appeal_issues_hash_array = issues[hearing.appeal_vacols_id] || []
+        appeal = Appeal.find_or_initialize_by(vacols_id: hearing.appeal_vacols_id)
```
Comment (aroltsch) [inline app/repositories/hearing_repository.rb]: You probably want to call find_or_create_by. `initialize` will not save it in the DB

File: app/models/vacols/case_hearing.rb [hunk 7]
```diff
@@ -77,7 +77,7 @@ def select_hearings
              :board_member, :mduser,
              :mdtime, :sattyid,
              :bfregoff, :bfso,
-             :bfcorkey, :bfddec, :bfdc,
+             :bfcorkey, :bfddec, :bfdc, :bfkey,
```
Comment (aroltsch) [inline app/models/vacols/case_hearing.rb]: I think you can use `folder_nr` because it should be the same as `bfkey`

File: app/models/appeal.rb [hunk 8]
```diff
@@ -418,6 +418,8 @@ def worksheet_issues
     super
   end
 
+  delegate :count, to: :worksheet_issues, prefix: true
```
Comment (aroltsch) [inline app/models/appeal.rb]: So the count will reflect even the issues they added? Is that what we want?

File: app/repositories/hearing_repository.rb [hunk 9]
```diff
@@ -8,7 +8,22 @@ def upcoming_hearings_for_judge(css_id)
         VACOLS::CaseHearing.upcoming_for_judge(css_id) +
           VACOLS::TravelBoardSchedule.upcoming_for_judge(css_id)
       end
-      hearings_for(MasterRecordHelper.remove_master_records_with_children(records))
+      hearings = hearings_for(MasterRecordHelper.remove_master_records_with_children(records))
+      non_master_record_hearings = hearings.select { |h| h.master_record == false }
```
Comment (aroltsch) [inline app/repositories/hearing_repository.rb]: You can remove this line. You can do the following: ```hearings.map(&:appeal_vacols_id).compact``` which will remove `nil` values and master records should have `nil` for `appeal_vacols_id`. Please double check that though.

File: app/repositories/hearing_repository.rb (hunk 6 above)
Comment (aroltsch) [reply app/repositories/hearing_repository.rb]: Also theoretically the appeal should have been already created: https://github.com/department-of-veterans-affairs/caseflow/blob/06bfe6f45ecc6ec42ce732905b411f9aec8da8fb/app/models/hearing.rb#L180

File: app/repositories/hearing_repository.rb [hunk 10]
```diff
@@ -8,7 +8,22 @@ def upcoming_hearings_for_judge(css_id)
...
+      # To speed up the daily docket and the hearing worksheet page loads, we pull in issues for appeals here.
+      Hearing.repository.load_issues(non_master_record_hearings)
```
Comment (aroltsch) [inline app/repositories/hearing_repository.rb]: You can probably remove this `Hearing.repository`

File: app/models/appeal.rb (hunk 8 above)
Comment (sharonwarner) [reply app/models/appeal.rb]: Correct.

**Pull request merged**